src/display/editor/ink.js

```diff
@@ -0,0 +1,689 @@
+   */
+  #redraw() {
+    this.#setStroke();
+
+    if (this.isEmpty()) {
+      this.#updateTransform();
+      return;
+    }
+
+    const { ctx, height, width } = this;
+    ctx.setTransform(1, 0, 0, 1, 0, 0);
+    ctx.clearRect(0, 0, width, height);
+    this.#updateTransform();
+    for (const path of this.bezierPath2D) {
+      ctx.stroke(path);
+    }
+  }
+
+  /**
+   * Commit the curves we have in this editor.
+   * @returns {undefined}
+   */
+  commit() {
+    if (this.#disableEditing) {
+      return;
+    }
+    this.disableEditMode();
+
+    this.#disableEditing = true;
+    this.div.classList.add("disabled");
+
+    this.#fitToContent();
+  }
+
+  /** @inheritdoc */
+  focusin(/* event */) {
+    super.focusin();
+    this.enableEditMode();
+  }
+
+  /**
+   * onmousedown callback for the canvas we're drawing on.
+   * @param {MouseEvent} event
+   * @returns {undefined}
+   */
+  canvasMousedown(event) {
+    if (!this.isInEditMode() || this.#disableEditing) {
+      return;
+    }
+
+    event.stopPropagation();
+
+    this.canvas.addEventListener("mouseleave", this.#boundCanvasMouseleave);
+    this.canvas.addEventListener("mousemove", this.#boundCanvasMousemove);
+
+    this.#startDrawing(event.offsetX, event.offsetY);
+  }
+
+  /**
+   * onmousemove callback for the canvas we're drawing on.
+   * @param {MouseEvent} event
+   * @returns {undefined}
+   */
+  canvasMousemove(event) {
+    event.stopPropagation();
+    this.#draw(event.offsetX, event.offsetY);
+  }
+
+  /**
+   * onmouseup callback for the canvas we're drawing on.
+   * @param {MouseEvent} event
+   * @returns {undefined}
+   */
+  canvasMouseup(event) {
+    if (this.isInEditMode() && this.currentPath.length !== 0) {
+      event.stopPropagation();
+      this.#endDrawing(event);
+    }
+  }
+
+  /**
+   * onmouseleave callback for the canvas we're drawing on.
+   * @param {MouseEvent} event
+   * @returns {undefined}
+   */
+  canvasMouseleave(event) {
+    this.#endDrawing(event);
+  }
+
+  /**
+   * End the drawing.
+   * @param {MouseEvent} event
+   */
+  #endDrawing(event) {
+    this.#stopDrawing(event.offsetX, event.offsetY);
+
+    this.canvas.removeEventListener("mouseleave", this.#boundCanvasMouseleave);
+    this.canvas.removeEventListener("mousemove", this.#boundCanvasMousemove);
+  }
+
+  /** @inheritdoc */
+  render() {
+    if (this.div) {
+      return this.div;
+    }
+
+    super.render();
+    this.canvas = document.createElement("canvas");
+    this.canvas.className = "inkEditorCanvas";
+    this.div.appendChild(this.canvas);
+    this.ctx = this.canvas.getContext("2d");
+
+    this.div.classList.add("editing");
+
+    if (this.width) {
+      // This editor has been created in using copy (ctrl+c).
+      this.setAt(this.x + this.width, this.y + this.height);
+      this.setDims(this.width, this.height);
+      this.#setCanvasDims();
+      this.#redraw();
+    }
+
+    const observer = new ResizeObserver(entries => {
```

Comment:
*This does ___not___ require any changes here, however:*

According to https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver#browser_compatibility, this isn't available in all Safari versions that we currently "support".

Given first of all that this isn't even enabled by default and secondly that our Safari-support is listed as partial, this isn't a big issue as far as I'm concerned.
Hence I'd suggest that we just add the following code:
```diff
diff --git a/web/app_options.js b/web/app_options.js
index 243aa9e11..d75277c56 100644
--- a/web/app_options.js
+++ b/web/app_options.js
@@ -38,6 +38,12 @@ if (typeof PDFJSDev === "undefined" || PDFJSDev.test("GENERIC")) {
       compatibilityParams.maxCanvasPixels = 5242880;
     }
   })();
+
+  (function checkResizeObserver() {
+    if (typeof ResizeObserver === "undefined") {
+      compatibilityParams.annotationEditorEnabled = false;
+    }
+  })();
 }
 
 const OptionKind = {
@@ -309,6 +315,9 @@ if (
     kind: OptionKind.VIEWER,
   };
 
+  defaultOptions.annotationEditorEnabled.compatibility =
+    compatibilityParams.annotationEditorEnabled;
+
   defaultOptions.renderer.kind += OptionKind.PREFERENCE;
 } else if (PDFJSDev.test("CHROME")) {
   defaultOptions.disableTelemetry = {

```